File: app/src/main/java/me/saket/dank/ui/subreddit/SubmissionThumbnailTypeMinusNsfw.java
package me.saket.dank.ui.subreddit;

import net.dean.jraw.models.Submission;

public enum SubmissionThumbnailTypeMinusNsfw {

  URL_STATIC_ICON,
  URL_REMOTE_THUMBNAIL,
  SELF_POST,
  NONE,
  UNKNOWN;

  /**
   * Like {@link Submission#getThumbnailType()}, but doesn't consider if a submission is NSFW, assuming that NSFW content is allowed.
   */
  public static SubmissionThumbnailTypeMinusNsfw parse(Submission submission) {
    //Timber.d("-------------------");
    //Timber.i("%s", submission.getTitle());
    //Timber.i("Thumb type: %s", submission.getThumbnailType());
    //Timber.i("Thumbnail: %s", submission.getThumbnail());
    //Timber.i("Thumbnails: %s", submission.getThumbnails());

    switch (submission.getThumbnailType()) {
      case NSFW:
        if (submission.isSelfPost()) {
          return SELF_POST;
        } else if (submission.getThumbnails() == null) {
          return URL_STATIC_ICON;
        } else {
          return URL_REMOTE_THUMBNAIL;
        }

      case DEFAULT:
        if (submission.getThumbnail() == null) {
          return NONE;
        } else {
          throw new AssertionError("Submission with DEFAULT thumbnail type and non-null thumbnail URL");
        }

      case SELF:
        return SELF_POST;

      case URL:
        if (submission.getThumbnails() == null) {
          return URL_STATIC_ICON;
        } else {
          return URL_REMOTE_THUMBNAIL;
        }

      case NONE:
        return NONE;

      default:
        return UNKNOWN;
    }
  }
}

File: app/src/main/java/me/saket/dank/ui/subreddit/SubmissionThumbnailTypeMinusNsfw.kt
package me.saket.dank.ui.subreddit

import net.dean.jraw.models.Submission
import net.dean.jraw.models.ThumbnailType

enum class SubmissionThumbnailTypeMinusNsfw {

  URL_STATIC_ICON,
  URL_REMOTE_THUMBNAIL,
  SELF_POST,
  NONE,
  UNKNOWN;

  companion object {
    /**
     * Like [Submission.getThumbnailType], but doesn't consider if a submission is NSFW, assuming that NSFW content is allowed.
     */
    fun parse(submission: Submission): SubmissionThumbnailTypeMinusNsfw {
      //Timber.d("-------------------");
      //Timber.i("%s", submission.getTitle());
      //Timber.i("Thumb type: %s", submission.getThumbnailType());
      //Timber.i("Thumbnail: %s", submission.getThumbnail());
      //Timber.i("Thumbnails: %s", submission.getThumbnails());

      return when (submission.thumbnailType) {
        ThumbnailType.NSFW -> when {
          submission.isSelfPost -> SELF_POST
          submission.preview == null -> URL_STATIC_ICON
          else -> URL_REMOTE_THUMBNAIL
        }
        ThumbnailType.DEFAULT -> NONE
        ThumbnailType.SELF -> SELF_POST
        ThumbnailType.URL -> when (submission.preview) {
          null -> URL_STATIC_ICON
          else -> URL_REMOTE_THUMBNAIL
        }
        ThumbnailType.NONE -> NONE
        else -> UNKNOWN
      }
    }
  }
}
